### DISCUSSION TITLE
Set up an Auth0 Provider, but my access token is ignored and validated against the local Jwt.js

### DISCUSSION THREAD

**Discussion by martaver** [strapi/strapi#7582] [Legacy Topics]
Hi everyone, new to Strapi here so I suspect I'm just missing something blindlingly obvious.

I'm using Strapi with the graphql plugin enabled.

My frontend is a nextjs app.

I want a single authentication provider across all of these services. Users login and register with Auth0 in the front-end app and then are able to access back-end services using the same credentials. I have the following set up in auth0:

**Applications** (see [Auth0 Application Types](https://auth0.com/docs/applications))

- app: the nextjs app, a regular web application, where the nodejs server takes a client secret, but its not exposed to client.
- cms: the strapi server, a machine-to-machine application, configured with grant 'Authorization code'.

**APIs** (see this [short description](https://auth0.com/docs/architecture-scenarios/spa-api/part-1#api-authentication-and-authorization))
- api: a single audience that includes my Application 'cms'.

My nextjs app has the 'app' Application credentials provided to it. Strapi has the 'cms' Application's credentials provided to it. My nextjs app's user session obtains an access_token for the audience that identifies 'api'.

I want users to be able to also connect to Strapi in order to load and edit content with the same session. User permissions are significant in my app, so Strapi needs to be aware of the user and protect unauthorised access.

Right now, my Auth0 provider in Strapi seems to work: I can use to `/connect/auth0` to log in with Auth0 and/or grant permissions for Strapi to access the user profile, and Strapi successfully creates a user entry with the provider type 'auth0'.

I'm failing when trying to authenticate a request for content though...

In my nextjs app, I log in using Auth0, obtain an access token for the audience of 'api', and use that access token as my auth header in my graphql request to Strapi, and this is the error I get:

```
{
  "errors": [
    {
      "message": "Invalid token.",
      "locations": [
        {
          "line": 3,
          "column": 11
        }
      ],
      "path": [
        "test"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "isBoom": true,
          "isServer": false,
          "data": null,
          "output": {
            "statusCode": 401,
            "payload": {
              "statusCode": 401,
              "error": "Unauthorized",
              "message": "Invalid token."
            },
            "headers": {}
          },
          "stacktrace": [
            "Error: Invalid token.",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-users-permissions/services/Jwt.js:58:27",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/jsonwebtoken/verify.js:121:14",
            "    at getSecret (/Users/martaver/Projects/sightful/git/sightful/node_modules/jsonwebtoken/verify.js:90:14)",
            "    at Object.module.exports [as verify] (/Users/martaver/Projects/sightful/git/sightful/node_modules/jsonwebtoken/verify.js:94:10)",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-users-permissions/services/Jwt.js:52:11",
            "    at new Promise (<anonymous>)",
            "    at Object.verify (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-users-permissions/services/Jwt.js:51:12)",
            "    at Object.getToken (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-users-permissions/services/Jwt.js:38:17)",
            "    at module.exports (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-users-permissions/config/policies/permissions.js:13:77)",
            "    at dispatch (/Users/martaver/Projects/sightful/git/sightful/node_modules/koa-compose/index.js:42:32)",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-utils/lib/policy.js:68:11",
            "    at dispatch (/Users/martaver/Projects/sightful/git/sightful/node_modules/koa-compose/index.js:42:32)",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/koa-compose/index.js:34:12",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/services/resolvers-builder.js:110:11",
            "    at field.resolve (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/graphql-extensions/dist/index.js:134:26)",
            "    at field.resolve (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/apollo-server-core/dist/utils/schemaInstrumentation.js:52:26)",
            "    at resolveFieldValueOrError (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/graphql/execution/execute.js:467:18)",
            "    at resolveField (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/graphql/execution/execute.js:434:16)",
            "    at executeFields (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/graphql/execution/execute.js:275:18)",
            "    at executeOperation (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/graphql/execution/execute.js:219:122)",
            "    at executeImpl (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/graphql/execution/execute.js:104:14)",
            "    at Object.execute (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/graphql/execution/execute.js:64:35)",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/apollo-server-core/dist/requestPipeline.js:249:48",
            "    at Generator.next (<anonymous>)",
            "    at /Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/apollo-server-core/dist/requestPipeline.js:8:71",
            "    at new Promise (<anonymous>)",
            "    at __awaiter (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/apollo-server-core/dist/requestPipeline.js:4:12)",
            "    at execute (/Users/martaver/Projects/sightful/git/sightful/node_modules/strapi-plugin-graphql/node_modules/apollo-server-core/dist/requestPipeline.js:228:20)"
          ]
        }
      }
    }
  ],
  "data": {
    "test": null
  }
}
```

So based on this stacktrace, I can see that the access token is being parsed correctly and that the error is being thrown when the access token is being verified against the server secret configured in `users-permissions/config/jwt.js`.

I'm not entirely sure if I understand the error, though... even if I provide the auth0 client secret for 'cms' by hardcoding it, the verification fails.

And either way, at any rate, shouldn't it be using the secret I've configured in my auth0 provider? Why is it falling back to the default local jwt provider?

At the same time, how would it know which provider to try to authenticate with? Does it try all of them that are enabled?

I'm scratching my head awfully vigorously over here...

Can anyone please help me crack this nut? What am I missing?

**Comment by derrickmehaffy** [comment]
Hi @martaver so a couple things I wanted to clarify:

- How are you integrating with Auth0 here, are you [adding a custom provider](https://strapi.io/documentation/v3.x/plugins/users-permissions.html#adding-a-new-provider-to-your-project) though Grant/Purest (it does support Auth0, we just don't have it included within Strapi by default): https://github.com/simov/grant.
- You mentioned: 

> I want a single authentication provider across all of these services. Users login and register with Auth0 in the front-end app and then are able to access back-end services using the same credentials.

By this you don't mean using the same token to access things like REST/GraphQL and the Strapi admin panel correct, because currently this isn't possible until we add support for SSO on the Admin. (See this card on our roadmap: https://portal.productboard.com/strapi/1-roadmap/c/61-sso-authentication-providers)

**Reply by martaver** [reply]
Heya, thanks for getting back to me!

I don't it necessarily has to be the SAME token... some kind of token exchange is perfectly fine. I just need to avoid the user having to log in more than once.

My goal is that the user logs in with the front-end once, through Auth0, and then is able to make authenticated requests to backend services, including the strapi cms. It's fine if the app has to do some kind of token exchange in order to make authenticated requests to backend services, and fine if the user has to confirm some extra permissions dialogs once. The back-end services need to be able to get the user profile so they can be identity aware. It's not okay if the user has to log in more than once.

That should be achievable with strapi, shouldn't it?

**Reply by derrickmehaffy** [reply]
Currently you won't be able to use Auth0 to authenticate into the Strapi admin panel, but the user end routes yes, you should just need to add the Auth0 Provider.

**Comment by martaver** [comment]
Okay, yeah I don't care about using Auth0 to log in to the admin panel -
just the end user auth :)

So I've added the Auth0 provider... but when I send a request token from
the front-end with a bearer auth containing an access token from the
front-end, strapi fails the auth as I described above. It's hard to tell
what's going on though... is the Auth0 provider failing auth first, and so
it falls back on the local JWT?

**Reply by martaver** [reply]
I actually don't think this is the case because no failed login attempts are logged in Auth0...

Can you give any insight as to why Strapi would be ignoring the Auth0 provider and attempting to validate using local JWT instead?

Do I have to implement a custom permissions policy to do this? If so, how would I get access to the provider and use it to validate?

**Reply by derrickmehaffy** [reply]
By adding in the Auth0 Provider you won't be using a token from Auth0 to actually make requests, it would still be a token that Strapi created, if you wanted to use an Auth0 token to make requests that is something you will have to do deep customization of the users-permissions plugin to accomplish. 

Maybe this guide can help you: https://strapi.io/documentation/v3.x/guides/jwt-validation.html#customize-the-jwt-validation-function

**Reply by martaver** [reply]
Yep seen that, that's why I was asking about the custom permissions policy! If I go to implement my own JWT verification though, what I don't understand is how I can detect the type of token received to pass it to the appropriate provider for validation. Can you shed some light on this?

**Reply by derrickmehaffy** [reply]
I don't have any information on this to provide as we don't really test nor implement that at all ourselves. I will ping @petersg83 or @alexandrebodin to see if they have any suggestions.
